packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js

```diff
@@ -1047,7 +1047,10 @@ export function canHydrateInstance(
   props: Props,
   inRootOrSingleton: boolean,
 ): null | Instance {
-  while (instance.nodeType === ELEMENT_NODE) {
+  while (
+    instance.nodeType === ELEMENT_NODE ||
+    (inRootOrSingleton && instance.nodeType === TEXT_NODE)
```

Comment:
This isn't quite right b/c the instance is refined to Element below and text nodes aren't elements. But I see what are you are trying to do. As commented below we're being somewhat conservative here but if you could make the case that there is widespread use of extensions/scripts that insert text nodes in the body before primary content then we can consider changing the heuristic

Reply:
Yeah so frameworks uses text nodes as a reference to a fragment because DocumentFragment [loses their children](https://github.com/whatwg/dom/issues/736) reference when appended in the dom. Or also the React portal adds text around portal nodes. 
Our use case is that pieces of the page like a header are rendered outside of React as inline scripts. We use suppressHydrationWarning for those DOM tree, but they also have portal similar implementation to render elements on the body causing the issue when the script is loaded.

example of React Portal rendered tree from the [documentation](https://react.dev/reference/react-dom/createPortal) as an example of the approach used
<img width="1050" alt="Screenshot 2024-09-30 at 4 44 59 PM" src="https://github.com/user-attachments/assets/d0e1c11b-5766-4d84-9010-a91904c908d1">